control-plane/connect-inject/webhook/consul_dataplane_sidecar.go

```diff
@@ -140,6 +154,54 @@ func (w *MeshWebhook) consulDataplaneSidecar(namespace corev1.Namespace, pod cor
 	return container, nil
 }
 
+// Configures graceful shut down for the sidecar.
+func (w *MeshWebhook) envoySidecarGracefulShutdown(pod corev1.Pod) (*corev1.Handler, error) {
+
+	grace, err := strconv.ParseBool(pod.Annotations[constants.AnnotationSidecarProxyGracefulShutdown])
+
+	if err != nil || !grace {
+		return nil, err
+	}
+
+	preStop := &corev1.Handler{
+		Exec: &corev1.ExecAction{
+			Command: []string{
+				"/bin/sh",
+				"-c",
+				"while [ $(netstat -plunt | grep tcp | grep -v envoy | grep -v consul-dataplane | wc -l) -ne 0 ]; do sleep 1; done",
+			},
+		},
+	}
+
+	return preStop, nil
+}
+
+// Ensures that the sidecar is the first container to start up.
+func (w *MeshWebhook) envoySidecarHoldApplicationUntilProxyStarts(pod corev1.Pod) (*corev1.Handler, error) {
+
+	hold, err := strconv.ParseBool(pod.Annotations[constants.AnnotationSidecarProxyHoldApplicationUntilProxyStarts])
+
+	if err != nil || !hold {
+		return nil, err
+	}
+	postStart := &corev1.Handler{
+		Exec: &corev1.ExecAction{
+			Command: []string{
+				"/bin/sh",
+				"-c",
+				`total_time=0; until wget --spider localhost:19000;` +
```

Comment:
The `/ready` endpoint should be more appropriate for this check IMHO:

```
#!/bin/sh

while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' localhost:19000/ready)" != "200" ]]; do
  sleep 1
done
```

Reply:
We do `while [[ "$(curl -sf http://127.0.0.2:19000/ready)" != "LIVE" ]]; do sleep 1; done` with great success :)